JK generic rotation

huang-kun opened the pull request:
I added a clockwise rotation button alongside with the default counterclockwise rotation button, which introduced a new clockwise rotation direction to TOCrop.

huang-kun commented:
This pull request is about adding a new rotation button directly into project. Here is a summary of changes:

1. Add a new rotation button to TOCropToolbar UI.
2. Add a new rotation API to TOCropView for providing both clockwise and counterclockwise rotation of image.
3. Change the fixed internal implementation of image rotation method by making it compatible with new rotation.
4. Encapsulate the TOCropToolbar button's frame calculation and make it more accurate and reusable for both horizontal and vertical toolbar layout.

These changes deserve a version update.

TimOliver commented:
Thanks a lot for that! I just incorporated those changes.

That being said, I did change a few things:

1. I made the clockwise button optional, and off by default. This is mainly since I don't think the majority of users (including me in my own use case) need two rotation buttons.
2. I moved the clockwise button to the far right of the panel, since I think separately those two icons looks slightly more visually confusing (The goal is their placement also helps indicate their direction)

Let me know if you need me to do anything else!

huang-kun commented:
Hi Tim


Some suggestions you may want to know from my point of view.


1. I modified the internal properties for adding new rotation. Each kind of rotation should be combined with a control (button) in this app and the only rotationButton from original TOCropToolbar seems not good. So I used two new properties to replace the old rotationButton and this change make more sense. Since the internal structure of TOCropToolbar has changed, for developers who visit the “_rotationButton” from outside of TOCropToolbar will definitely hit a crash because the original ivar didn’t exist any more. We may not be responsible for this kind of crash because the “_rotationButton” is not declared public, but I did change the public block property “rotateButtonTapped” to other two new block properties. You might want to also redeclare “rotateButtonTapped” property and make it’s setter and getter pointing to the default counterclockwise action so people whoever used “rotateButtonTapped" will not see no compile error. (OR NOT)


2. Have you (or other contributors) ever thought re-designign parts of TOCropViewController’s architecture that providing the ability for other developer to customize the layout of the default crop view and control ? For a lot of reasons and specific demands, we can not just use without customizing (e.g. make button’s image different or prefer the toolbar on the right side in landscape mode or maybe speed up the crop-zoom animation, etc). In my project, I also use DBCamera (https://github.com/danielebogo/DBCamera) to capture image. This camera project is an example that provides a way for developers to customize the default camera even subclass their camera view. I bet people can custom TOCropViewController from lots of approaches, however these are not official APIs and not introduced on README. (OR NOT if TOCrop is not designed that way from beginning)


3. At the end of Feb, I need to continue work on my project binding with this fantastic cropping tool and I got a multi-cropping gesture support request (Using two fingers to manipulate the crop box frame) which is not supported on TOCrop by now. Honestly I haven’t looked at the gesture codes in detail from TOCrop, but I will see what I can do for that. Any interesting ?


Original Message
Sender:Tim Olivernotifications@github.com
Recipient:TimOliver/TOCropViewControllerTOCropViewController@noreply.github.com
Cc:huang-kunjack-huang-developer@foxmail.com
Date:星期日, 1月 24, 2016 22:10
Subject:Re: [TOCropViewController] JK generic rotation (#45)


Thanks a lot for that! I just incorporated those changes.
That being said, I did change a few things:
I made the clockwise button optional, and off by default. This is mainly since I don't think the majority of users (including me in my own use case) need two rotation buttons. I moved the clockwise button to the far right of the panel, since I think separately those two icons looks slightly more visually confusing (The goal is their placement also helps indicate their direction)
Let me know if you need me to do anything else!
—
Reply to this email directly or view it on GitHub.
